src/Shared/Model/Providers/NuGetProvider.cs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Microsoft Corporation. Licensed under the MIT License.
+
+namespace Microsoft.CST.OpenSource.Model.Providers;
+
+using Contracts;
+using Metadata;
+using Microsoft.CST.OpenSource.PackageManagers;
+using NLog;
+using NuGet.Packaging.Core;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+using PackageUrl;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Utilities;
+
+public class NuGetProvider : IManagerProvider
+{
+    private readonly SourceCacheContext _sourceCacheContext = new();
+    private readonly SourceRepository _sourceRepository = Repository.Factory.GetCoreV3("https://api.nuget.org/v3/index.json");
+    private readonly NuGetLogger _logger = new(LogManager.GetCurrentClassLogger());
+
+    /// <summary>
+    /// Instantiates a new <see cref="NuGetProvider"/>.
+    /// </summary>
+    public NuGetProvider()
+    {
+    }
+
+    /// <inheritdoc />
+    public async Task<string?> DownloadAsync(
+        PackageURL packageUrl,
+        string topLevelDirectory,
+        string targetDirectory,
+        bool doExtract,
+        bool cached = false,
+        CancellationToken cancellationToken = default)
+    {
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        // Create a new memory stream to populate with the .nupkg.
+        using MemoryStream packageStream = new();
```

Comment:
How large can NuGet package files get? Have we considered something like downloading to a temporary file instead of writing to RAM?

Reply:
Looks like there are some decent ways to go about using a temporary file: https://stackoverflow.com/questions/34776031/how-to-create-a-temp-file-that-is-automatically-deleted-after-the-program-termin

Reply:
This point could easily be addressed in a separate PR though - not a blocker here.

Reply:
Probably will address in another PR. But this implementation is straight from the documentation just as an FYI:
https://docs.microsoft.com/en-us/nuget/reference/nuget-client-sdk#download-a-package

Reply:
Addressed.